### PR TITLE
[WIP] added guide url to the commands in kubectl

### LINKED ISSUE #31267: Improve kubectl error messages
List here:
https://github.com/kubernetes/community/wiki/Roadmap:-kubectl

Cross them out as you complete them

---
## List of additional issues
- [ ] #34457  exec pods on master unsupported
- [ ] #30081 --dry-run + -i unsupported

### PR DESCRIPTION
**Pull request by dgonzalez**
Do not merge. Work in progress. It will partially fix: https://github.com/kubernetes/kubernetes/issues/31267. Please @AdoHe have a look to this. Totes simple and I will keep adding code to fix other commands.

In order to show the changes, execute: `kubectl get` and along all the valid options will link you to the guide in the docs.

This is a point in the roadmap for kubectl. Once I get the green on format and procedure I will add the same for all the other commands (like annotate, expose...)

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33435)

### CODE REVIEW

**Comment by dgonzalez** [comment]
I will squash the commits once we are happy with them.

**Comment by AdoHe** [comment]
@dgonzalez could you please link the specific issue here, #31267 is a quite big one.

**Comment by dgonzalez** [comment]
If you see the roadmap for kubectl:
- https://github.com/kubernetes/community/wiki/Roadmap:-kubectl

There is a section `Improve help / error messages / output` and inside `Link to docs (kubernetes.io/docs)`.

What I understand from it is adding the links to the docs in the kubectl output and I thought the best place was the help of the commads:

`kubectl get --help` -> it will print the help for get and also link you to the guide on how to use it.

It is still work in progress.

Should I create a smaller issue just for this? As you said... it is quite a big one.

Thanks

File: pkg/kubectl/cmd/explain.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2014 The Kubernetes Authors.
+Copyright 2016 The Kubernetes Authors.
```
Comment (deads2k) [inline pkg/kubectl/cmd/explain.go]: This date stays the same.  It's the date the file was created, not the date of the last modification.

**Comment by deads2k** [comment]
@fabianofranz We get to choose factorization here.  Do we want it natively substitutable?  If so, you should probably provide a sample.

File: pkg/kubectl/cmd/explain.go (hunk 1 above)
Comment (dgonzalez) [inline pkg/kubectl/cmd/explain.go]: Yep. I am reverting it.

**Comment by dgonzalez** [comment]
@deads2k you mean something somehow generic that will fit all the commands? I think I didn't fully get you.

**Comment by deads2k** [comment]
> @deads2k you mean something somehow generic that will fit all the commands? I think I didn't fully get you.

An argument passed to a string format.  I'd really like to have to a way to check for link-rot automatically.

**Comment by fabianofranz** [comment]
@deads2k we definitely want it substitutable, but we already have the issue - "kubectl" is already hardcoded in the examples and some long descriptions. Should we try to fix that too once we are here?

We could either start by passing more string args to the cmd constructors

```
...
create_example = dedent.Dedent(`
  # Create a pod based on the JSON passed into stdin.
  cat pod.json | %[1]s %[2]s -f -
  For more info, please visit: %[3]s`)

func NewCmdCreate(baseCmd,  cmdName, cmdDetailsURL string, f *cmdutil.Factory, out io.Writer) *cobra.Command {
...
  Example: fmt.Sprintf(create_example, baseCmd, cmdName, cmdDetailsURL)
...
```

or establish an entity at the `pkg/kubectl/cmd` level to hold it?

```
type CmdCustomizations {
  baseCmd       string
  cmdName       string 
  cmdDetailsURL string
}

...
func NewCmdCreate(f *cmdutil.Factory, out io.Writer, custom CmdCustomizations) *cobra.Command {
```

**Comment by dgonzalez** [comment]
I'll be up for the `cmdCustomizations`. It looks cleaner to me.

**Comment by dgonzalez** [comment]
Btw, if we go this way it is going to be a big refactor... we probably should split this issue from the original one.

**Comment by deads2k** [comment]
I like the idea of a separate context struct that holds the "normal" stuff that every command needs (minus the factory).  So maybe even stdout and stderr to make sure all commands are accepting them.

**Comment by fabianofranz** [comment]
Yeah. It could be called `CmdContext` and reside in `pkg/kubectl/cmd/context.go`.

**Comment by AdoHe** [comment]
yes, this sounds reasonable to me. @fabianofranz we need to open a separate a PR to track this, and we can hold this PR until we get `CmdContext`. wdyt?

**Comment by dgonzalez** [comment]
I will work on the `CmdContext` if you don't mind and then I will come back to this one.

**Comment by fabianofranz** [comment]
@AdoHe sure, works for me. Thanks @dgonzalez!

**Comment by dgonzalez** [comment]
I will close this one once we follow up in https://github.com/kubernetes/kubernetes/pull/34418/files

**Pull request closed without merging**